### ISSUE TITLE
Masking support

### ISSUE THREAD

**Issue by SimonHeybrock** [scipp/scipp#217]
Masking and ROI (region-of-interest) selection is frequently used in neutron data reduction. We have previously demonstrated that a crude way of supporting this is to simply store masks as labels, data, or attributes. This can certainly work, but it is not clear whether this is convenient and intuitive. We have to consider the following cases:

- More than one mask.
- Masking should be reversible, i.e., not touch the data.
- Mask views may be useful.
- Plotting must be able to include masks.
- In binary operations that operate on each element we need to combine the masks somehow. For example, when adding two datasets we may want to `or` the masks (for `bool` masks, what about `float` masks?).
- For some operations masking is not reversible, for example `sum`. These will need to apply masks, but we may require user-control.

One possible solution would be to introduce an extra category `masks`, in addition to `coords`, `labels`, and `attrs`. This would likely be intuitive, and would give a natural way to implement special handling for masks (`or` operation, if binary operation applied to data). This is then just a continuation of having different "sections" of a data array or dataset implementing different handling:

- coords and labels are *compared*.
- data is operated on (`+-*/`, ...)
- masks are `OR`ed
- attributes are merged or dropped

UX/UI:

```python
# add mask
dataset.masks['roi'] = ...
# remove mask
del dataset.masks['roi']
# include mask in plot
sc.plot(dataset, masks=['dead_pixels'])
# apply mask
dataset *= dataset.masks['dead_pixels']
```

**Comment by OwenArnold** [comment]
I like the suggesting of supporting `masks` as a new catagory, mainly because it will be more intuitive and operations involving masking should (OR) the masks.

Another feature for masking might be to "invert" them to create a ROI from a Mask and vica-versa. A potential use of this would be in allowing users to treat different regions of the same data independently.

For spacial masking, we need to be able to easily apply masking based on some simple criteria. i.e. if  `-0.1 <= x > 2.0`. This is possible in Mantid via `ComponentInfo()`/`DetectorInfo` and this has proved very useful, for some neutron scattering techniques, this functionality would be critical.

There are also considerations for "bin masking". While discussing Mantid, note that there were recently some complications when bin masking was considered for histogram event mode data. I don't know what the conclusion of that was, but I'll investigate.

**Comment by SimonHeybrock** [comment]
> Another feature for masking might be to "invert" them to create a ROI from a Mask and vica-versa. A potential use of this would be in allowing users to treat different regions of the same data independently.

I am not sure what you mean here? I do not see a need for distinguishing masks and ROIs (but maybe there is?). Logic operations will of course be provided, and thanks to `transform` this is almost a 1-liner.

> For spacial masking, we need to be able to easily apply masking based on some simple criteria. i.e. if -0.1 <= x > 2.0. This is possible in Mantid via ComponentInfo()/DetectorInfo and this has proved very useful, for some neutron scattering techniques, this functionality would be critical.

I was not aware that this has been added to `DetectorInfo`? Is it like `numpy.where`? Scipp can do this using `transform`, something like:

```cpp
auto mask = transform<Eigen::Vector3d>(positions, [](const auto &pos) { return abs(pos[0]) < 0.1; });
```
... unfortunately not directly available in Python.

> There are also considerations for "bin masking".

Scipp handles everything in the same way, we can avoid the headaches in Mantid from having two kinds of masking. We can mask pixels using a 1D mask with `Dim.Position`, mask bins in all spectra using a 1D mask with `Dim.Tof`, or mask specific bins in specific pixels with a 2D mask depending on both dims.

**Comment by OwenArnold** [comment]
> I am not sure what you mean here? I do not see a need for distinguishing masks and ROIs 

I don't think there is, at least not as a storage requirement. This is good as we can store masks as Bool.

> I was not aware that this has been added to DetectorInfo

Not it hasn't been added like that, you still have to manually iterate, though there is masking support via iterators read/write.

**Comment by SimonHeybrock** [comment]
**Notes for implementer:**

I think all required changes are relatively straightforward and simple (mostly based on existing features or templates). However, there is a relatively large number of things to add, so I would recommend splitting this into multiple tasks:

1. Add mask support in `Dataset` and `DataArray`: This should mostly be copy and paste and can use the templated `ConstProxy` and `MutableProxy`:
   ```cpp
   using MasksConstProxy = ConstProxy<ProxyId::Coords, std::string>;
   using MasksProxy = MutableProxy<CoordsConstProxy>;
   ```
   See https://github.com/scipp/scipp/blob/64fef890dcbe9c4342ae7370d511c8675654bdbd/core/include/scipp/core/dataset.h#L35-L46.
   - Add map to store masks, similar to labels.
   - Add `masks()` method, similar to `labels()`, same for similar methods. This needs to be done for all proxies as well (`Dataset`, `DatasetConstProxy`, `DatasetProxy`, `DataArray`, `DataConstProxy`, `DataProxy`).
   - There is no need for sparse masks.
   - Implement ORing masks in binary operations, e.g., this https://github.com/scipp/scipp/blob/64fef890dcbe9c4342ae7370d511c8675654bdbd/core/dataset_binary_arithmetic.cpp#L19-L23 would need to be extended with `masks() |= other.masks()` (or rather something that takes the union of the present masks and ORs those that are present in both arguments; make sure to have a look at the `union_` function which does something similar but different for coordinates).
2. Basic boolean operators (`|=`, `|`, ...) can be implemented using `transform` and `transform_in_place`.
  - No need to implement more than needed initially. We might only need `OR` for now.
  - The only subtlety is handling the unit. Anything that is not `dimensionless` should cause a failure, I guess.
3. Python bindings.

**Comment by SimonHeybrock** [comment]
Implemented in #589.
